Add support for EDAlias as a case in SwitchProducer

makortel opened the pull request:
In a heterogeneous workflow it may happen that the (legacy) CPU and non-CPU modules are naturally defined so differently that the final legacy CPU products are produced in a very asymmetric way (e.g.
* CPU: producer `A` produces `a1`, `a2`, producer `B` produces `b1
* non-CPU: producer `D` produces `a1`, producer `C` produces `a2` and `b1`

). In order to be able to handle such cases flexibly and efficiently, this PR extends the SwitchProducer mechanism to allow EDAliases as cases (in addition to EDProducers). This required first some fixes to EDAliases:
* Throw if an EDAlias defines two instances pointing to the same real product
   * Currently the OutputModule will complain if an alias and the alias-for branches are attempted to be kept. This check does not work with SwitchProducer, whose output branches are not exactly alias branches (even if they're very close to). The premise is that it does not make sense to have both the alias and alias-for branches.
* If a job has an unscheduled EDProducer that registers a callback for new product registration such that the callback calls `produces()`, and that callback gets triggered by EDAlias (output) BranchDescriptions, the EDProducer BranchDescriptions have an incorrect on-demand status leading to a wrong ProductResolver and eventually a job to hang because of starvation in the TBB scheduler. This situation happens because the on-demand status for EDProducer BranchDescriptions is set before the EDAliases are processed in the Schedule. The fix is to move the setting of the on-demand status of BranchDescriptions to be done after the EDAlias output BranchDescriptions have been created.
  * The setting is also a bit more efficient now that it is done once instead of once per stream
* Made the python `EDAlias` cloneable and to inherit from `_Parameterizable`
  * Latter reduces copy-paste and enables `Modifier.toReplaceWith()`

Adding the support for EDAlias cases in SwitchProducer required the following changes
* SwitchProducer ProductResolvers can be set only after all EDAlias ProductResolvers have been set
* Added a new (intermediate) base class `DataManagingOrAliasProductResolver`, from which the `DataManaging`, `Alias`, and `SwitchBase` ProductResolvers inherit from, in order to be able use both `DataManaging` and `Alias` ProductResolvers as the `realProduct` in `SwitchBaseProductResolver`
* Added a special treatment for SwitchProducer in `ProductSelector` to detect the case when one attempts to keep in OutputModule both the real product, and the EDAlias-within-SwitchProducer. Similar check is already done for regular EDAliases (as mentioned above)
  * The same check does not work for SwitchProducers because there the `BranchID` of the SwitchProducer and the case product/alias are different, while in case of EDAlias the `BranchID` of the alias and the alias-for product as the same.
* In python `Process`, separate `SwitchProducers` from the normal `EDProducers`. This change is needed in order to append elements to both `@all_modules` and `@all_aliases` `vstring`s when filling the ProcessDescription. This way the case-EDAlias can away from the `Process.aliases_()` (as the case-EDProducers are not in `Process.producers_()`).
* As an additional sanity check, mark all case BranchDescriptions as transient in order to ignore them in OutputModule keep statements like `keep *_foo*_*_*`.
  * I also tried to "implicitly drop" all branches with '@' in `ProductSelector`, but that lead to SwitchProducer branches being dropped on inputs as they would be interpreted as branches depending on a dropped branch.

This PR also adds a generic mechanism to allow customized configuration entities to pass import statements to configuration dump. Related, currently the case-producer fragments end up in the configuration dump while they should not, this PR fixes that (as a byproduct).


Tested in CMSSW_10_5_X_2019-02-08-1100, no changes expected in current workflows.

makortel committed:
Allow SwitchProducer to have EDAlias case

- In configuration need some special cases for EDAlias
- Need to process Switch aliases after all EDAliases in Principal,
  otherwise the aliased-for ProductResolvers would not necessarily exist
- Add DataManagingOrAliasProductResolver to be able to connect
  SwitchBaseProductResolver to AliasProductResolver
- Extend the duplicate product check in ProductSelector to disallow
  keeping both a real product, and EDAlias-within-SwitchProducer alias
  of the product
- Add a test to verify that having a producer and
  SwitchProducer-with-EDAlias pointing to that producer in the wrong
  oder in a Path leads to an exception

makortel committed:
Make EDAlias to inherit from _Parameterizable, and add clone()

The clone() is needed to make SwitchProducer (that is clonable) to
support EDAliases as cases. Making EDAlias to inherit from
_Parameterizable reduces copy-paste.

Also add unit tests for using Modifier for EDAlias.

makortel committed:
Reduce copy-paste in SwitchProducer TestProcessor test

makortel committed:
Prevent EDAlias defining two instances pointing to the same product

Also add a test for the other kind of duplicate check

makortel committed:
Fix the on-demand status of BranchDescriptions created when new products are registered in a callback triggered by an EDAlias.

All BranchDescriptions have been created after processEDAliases(), so
let's move the call to BranchDescription::setOnDemand() there.

makortel committed:
Mark SwitchProducer case branches as transient so that they are always ignored in output

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25907/8386

- This PR adds an extra 156KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

DataFormats/Provenance
FWCore/Framework
FWCore/Integration
FWCore/ParameterSet


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @rovere, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on FWCore/Framework/src/Schedule.cc:
```diff
@@ -620,7 +644,38 @@ namespace edm {
+
+    // At this point all BranchDescriptions are created. Mark now the
+    // ones of unscheduled workers to be on-demand.
+    // TODO: what to do for alias branches???
```
What is the plan to resolve this TODO?

makortel replied:
Thanks for catching, the comment itself is a leftover. The alias branches are already dealt with below on lines 667-676.

cmsbuild commented:
-1

Tested at: 194c177eaf323e17a8c7d76727a9a209c7690fe0

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33082/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency**

* **Build**:

I found compilation error when building: 

<pre>Entering library rule at Mixing/Base
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-11-1100/src/Mixing/Base/src/PileUp.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-11-1100/src/Mixing/Base/src/BMixingModule.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-11-1100/src/Mixing/Base/src/SecondaryEventProvider.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-11-1100/src/Mixing/Base/src/SecondaryEventProvider.cc: In constructor 'edm::SecondaryEventProvider::SecondaryEventProvider(std::vector<edm::ParameterSet>&, edm::ProductRegistry&, std::shared_ptr<edm::ProcessConfiguration>)':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-11-1100/src/Mixing/Base/src/SecondaryEventProvider.cc:27:22: error: 'class edm::WorkerManager' has no member named 'setOnDemandProducts'
       workerManager_.setOnDemandProducts(preg, unscheduledLabels);
                      ^~~~~~~~~~~~~~~~~~~
gmake: *** [tmp/slc7_amd64_gcc700/src/Mixing/Base/src/MixingBase/SecondaryEventProvider.cc.o] Error 1
>> Building shared library tmp/slc7_amd64_gcc700/src/Mixing/Base/src/MixingBase/libMixingBase.so
c++: error: tmp/slc7_amd64_gcc700/src/Mixing/Base/src/MixingBase/SecondaryEventProvider.cc.o: No such file or directory
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented on FWCore/Integration/test/run_TestEDAlias.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+test=testEDAlias
```
How is this test run? I didn't notice anything in BuildFile.xml.

makortel replied:
Sigh, I forgot to add it again to BuildFile.xml, thanks for catching.

makortel committed:
Add a check that aliases to non-existing products fail only when the corresponding product is accessed

makortel committed:
Really run EDAlias test

makortel committed:
Remove obsolete comment

makortel committed:
Move onDemand-status setting from Schedule to ProductRegistry

ProductRegistry seems to be a natural place for the function (that is
needed also outside of FWCore).

makortel committed:
Migrate SecondaryEventProvider to ProductRegistry::setUnscheduledProducts()

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Addressed @Dr15Jones' comments, fixed the compilation failure, and added a check for an additional corner case.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25907/8387

- This PR adds an extra 244KB to repository

cmsbuild commented:
Pull request #25907 was updated. @cmsbuild, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 527a1c3cc88106ec3320d261d835a4f86b37dd19

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33085/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testGroupSelection had ERRORS
---> test testConditionDatabase_2 had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test testRecoMETMETProducers had ERRORS
---> test runtestTqafExamples had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test testTauEmbeddingProducers had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test testPhase2PixelNtuple had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test SiStripDAQ_O2O_test had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33085/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3097440
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3097242
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 527a1c3cc88106ec3320d261d835a4f86b37dd19

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33087/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testGroupSelection had ERRORS
---> test testTauEmbeddingProducers had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test testPhase2PixelNtuple had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestTqafExamples had ERRORS
---> test testRecoMETMETProducers had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test TestPoolInput had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test SiStripDAQ_O2O_test had ERRORS

cmsbuild commented:
Comparison job queued.

makortel commented:
At least the `testGroupSelection` works fine locally on CMSSW_10_5_X_2019-02-08-1100.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33087/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3097440
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3097242
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

makortel commented:
@cmsbuild, please test

Let's see if a new IB would make any difference.

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
as a general comment, it would be very nice to have this in pre2 ...

cmsbuild commented:
-1

Tested at: 527a1c3cc88106ec3320d261d835a4f86b37dd19

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33107/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test SiStripDAQ_O2O_test had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
250202.181 step2
<pre>runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/step2_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

makortel commented:
Only one unit test failure?
```
%MSG-w SiStripCabling:  SiStripO2OBadStrip:siStripO2O 13-Feb-2019 19:44:03 CET  Run: 1 Event: 1
[FedChannelConnection::apvPairNumber] Unexpected number of APV pairs: 65535
%MSG


[2019-02-13 19:44:03,856] INFO: @@@CMSSW job return code = -11@@@
Traceback (most recent call last):
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 181, in <module>
    main()
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 174, in main
    runjob(args)
  File "/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-13-1100/bin/slc7_amd64_gcc700/SiStripDAQPopCon.py", line 89, in runjob
    raise RuntimeError('O2O job FAILED!')
RuntimeError: O2O job FAILED!
Failure running SiStripDAQPopCon.py: status 1
```

The error in 250202.181 step2 is essentially a file read error.

makortel commented:
@cmsbuild, please test

I want to see how well the behaviour reproduces...

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 527a1c3cc88106ec3320d261d835a4f86b37dd19

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33121/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test SiStripDAQ_O2O_test had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33121/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3097440
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3097242
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

makortel committed:
Allow customized components outside FWCore to pass import statements to config dump

Achieved by creating a registry where these components can register
their import statements.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel committed:
Add a test to ensure that the case-producers do not 'leak' to the config dump

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25907/8432

- This PR adds an extra 252KB to repository

makortel commented:
Added two things for configuration
* unit test that the case-producer configuration fragments (with `@` in the module label) do not end up in the configuration dump
* a generic mechanism to allow customized configuration entities to pass import statements to configuration dump
  * this is a bit of feature creep for this PR, but technically the test of this feature conflicts with the previous test addition, and overall this PR is needed to remove the case-producer fragments from the dump

cmsbuild commented:
Pull request #25907 was updated. @cmsbuild, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

cmsbuild commented:
-1

Tested at: 527a1c3cc88106ec3320d261d835a4f86b37dd19

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33133/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test SiStripDAQ_O2O_test had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33133/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3097440
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3097242
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

makortel commented:
The error in runtestPhysicsToolsPatAlgos is again one of the weird ones
```
#4  <signal handler called>
#5  0x00007f03cbcb782c in std::_Rb_tree<edm::TypeIDBase, std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> >, std::_Select1st<std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> > >, std::less<edm::TypeIDBase>, std::allocator<std::pair<edm::TypeIDBase const, std::shared_ptr<edm::serviceregistry::ServiceWrapperBase> > > >::find(edm::TypeIDBase const&) () from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-14-1100/lib/slc7_amd64_gcc700/libFWCoreFramework.so
#6  0x00007f03cbf872e4 in edm::serviceregistry::ServicesManager::~ServicesManager() () from /cvmfs/cms-ib.cern.ch/nweek-02563/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_X_2019-02-13-2300/lib/slc7_amd64_gcc700/libFWCoreServiceRegistry.so
#7  0x00007f03cbf87421 in edm::serviceregistry::ServicesManager::~ServicesManager() () from /cvmfs/cms-ib.cern.ch/nweek-02563/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_X_2019-02-13-2300/lib/slc7_amd64_gcc700/libFWCoreServiceRegistry.so
#8  0x0000000000413786 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() ()
#9  0x00007f03cbc5db8a in edm::EventProcessor::~EventProcessor() () from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-02-14-1100/lib/slc7_amd64_gcc700/libFWCoreFramework.so
#10 0x000000000040e6e3 in (anonymous namespace)::EventProcessorWithSentry::~EventProcessorWithSentry() ()
#11 0x000000000040e17f in main ()

Current Modules:

Module: none (crashed)
```
happening at the very end of the job.

Dr15Jones commented:
@makortel please document in the comments what tests you've run to try to find the cause of the unit test failures (e.g. what tools you ran). Once documented, I will sign off on the pull request assuming the failures are unrelated.

makortel commented:
I ran all the failed tests in https://github.com/cms-sw/cmssw/pull/25907#issuecomment-462590189 in a local area in my development IB and in the IB used in the PR tests, all of those tests succeeded.

In #25914 I ran runtestRecoEgammaPhotonIdentification with valgrind, but got nothing relevant.

I'm still planning to play with ASAN for some of the tests in this PR. I'll anyway document all findings in this PR.

Dr15Jones commented:
+1
The unit test failures are also appearing in other pull requests as well.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 95d0376179b11503b5a4d171f4ec399517e67d92

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33151/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test testPhase2PixelNtuple had ERRORS
---> test TestPoolInput had ERRORS
---> test testRecoMETMETProducers had ERRORS
---> test runtestTqafExamples had ERRORS
---> test SiStripDAQ_O2O_test had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33151/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3098286
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3098088
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

makortel commented:
I tested the following tests in CMSSW_10_5_ASAN_X_2019-02-08-2300 (didn't have time to run all tests until this ASAN IB expired)

```
---> test testTauEmbeddingProducers had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test testPhase2PixelNtuple had ERRORS
---> test runtestTqafExamples had ERRORS
```

Dr15Jones commented:
@civanch @mdhildreth please review. We believe the test failures are from problems in the pull request testing system when too many tests have to be run and are NOT from this pull request.

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

civanch commented:
@makortel , for me it is a bit strange that this PR trigger problems of these tests. May it be that the modification makes the tests to be sensitive and they cannot work in parallel at the same node?

makortel commented:
@civanch I fully agree that the unit test failures are strange. But the failures are somewhat random (e.g. in https://github.com/cms-sw/cmssw/pull/25907#issuecomment-463367248 only `SiStripDAQ_O2O_test` failed, and 250202.181 because of file open errors), occur also in other PRs, and are not reproducible in local tests. 

Also @smuzaffar commented in https://github.com/cms-sw/cmssw/pull/25917#issuecomment-464771140
> We have noticed that when PR include too many packages (e.g. in this case over 700) then unit tests misbehave and are mostly killed by `timeout`

fabiocos commented:
I confirm that in my local test build I manage to execute successfully the failing tests. Let's try again...

cmsbuild commented:
-1

Tested at: 95d0376179b11503b5a4d171f4ec399517e67d92

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33193/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS
---> test runtestPhysicsToolsNanoAOD had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestRecoEgammaPhotonIdentification had ERRORS
---> test runtestRecoEgammaElectronIdentification had ERRORS
---> test SiStripDAQ_O2O_test had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25907/33193/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3098286
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3098088
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

boudoul commented:
Dear all, please note that the PR #25736 is also experiencing strange and random unit test failures ....

fabiocos commented:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25907/8490

- This PR adds an extra 256KB to repository

fabiocos commented:
@civanch @mdhildreth as far as I can see, the change under the simulation umbrella in Mixing/Base is just a direct consequence of the other updates (the method used has been replace by another one). The short matrix tests involving PU do not show any difference, and the random failures in the unit tests are not explicitly related to the SIM part, and looks most likely an issue with the test ssyetm itself (under investigation).

In agreement with @smuzaffar I will move forward with the merge of this PR, so as we may verify in an IB whether we observe similar problems (none so far).

fabiocos commented:
+1

fabiocos commented:
merge

to verify in an IB environment whether the observed random crashes appear as well, or whether the problem is related to the PR test environment

makortel commented:
Seems that all unit tests indeed succeeded in the IB.

fabiocos commented:
@makortel indeed this points towards an issue in the test system...
